kfp_notebook/pipeline/_notebook_op.py

```diff
@@ -236,3 +266,40 @@ def _artifact_list_to_str(self, pipeline_array):
                     ValueError("Illegal character ({}) found in filename '{}'.".format(INOUT_SEPARATOR, artifact_name))
             trimmed_artifact_list.append(artifact_name.strip())
         return INOUT_SEPARATOR.join(trimmed_artifact_list)
+
+    @staticmethod
+    def _normalize_label_value(value):
+
+        """Produce a Kubernetes-compliant label from value
+
+        Valid label values must be 63 characters or less and
+        must be empty or begin and end with an alphanumeric
+        character ([a-z0-9A-Z]) with dashes (-), underscores
+        (_), dots (.), and alphanumerics between.
+        """
+
+        if value is None or len(value) == 0:
+            return ''   # nothing to do
+
+        value = value[:63]  # enforce max length 63
+
+        # must begin with [0-9a-zA-Z]
+        valid_chars = string.ascii_letters + string.digits
+        if value[0] not in valid_chars:
+            value = 'a' + value[1:]
```

Comment:
Do we want to _replace_ the first character, rather than just prepend to it?  Since its arguably the most significant character, I'm thinking it should be preserved - albeit illegal as a label value.

Of course, the truncation would need to happen after the first-character check.

Reply:
Let's use example `#1234567890...YZ`
 - prepend `a` (`a#1234567890...YZ`) and
 - truncate if new total length exceeds the max 63 `a#1234567890...Y` OR
 - always truncate to retain original length `a#1234567890...Y`
WDYT?

Reply:
I guess in the spirit of your initial comment it would make most sense to preserve the original string as much of possible and only truncate if the length constraint is violated

Reply:
yes.  Only truncate to 63 if the _current_ length (i.e., after possible prefixing) is > 63.